src/Serializers/BaseSerializer.ts

```diff
@@ -110,18 +120,23 @@ export abstract class BaseSerializer<S, T extends IModelData, R = T> implements
     )(data);
 
     // Build the model's data
-    const modelData: T = flow(
+    const normalizeFieldPromises = flow(
       pick(fieldKeys),
-      mapValuesWithKeys(this.normalizeField(data)),
-    )(this.fields) as T;
+      mapWithKeys(await this.normalizeField(data)),
+    )(this.fields) as Promise<[string, string]>[];
+
+    // promise resolves with an array of tuples in the format: [key, value][]
+    // fromPairs then converts the array of tuples into an object
+    const pairs = await Promise.all(normalizeFieldPromises);
+    const modelData: T = fromPairs(pairs) as any;
```

Comment:
Why are we typing this value just to "untype" it again?

Reply:
The issue is the shape of the typed output from the `fromPairs` function. The function does take a generic:

```
fromPairs<T>(
  pairs: List<[PropertyName, T]> | null | undefined
): Dictionary<T>;
```

Unfortunately, I think they made an assumption on the shape of the output that is not valid in our case, given that this is the type definition of `Dictionary<T>`:

```
interface Dictionary<T> {
  [index: string]: T;
}
```

So, if we pass our `T extends IModelData`, we get back a `Dictionary<T>`, which kind of implies it's giving us back a map of `T`, not just `T` itself. We would need to find a way to just pass the values (minus the keys) of `T` for this to work... poorly. I tried simply casting the output to `T` instead of `any`, but TypeScript threw an error stating that `Dictionary<any> is not compatible with T`.

So, casting the output of `fromPairs` to `any`, allows us to then assign the `any` value onto the typed `modelData` as a `T`, so we could properly pass the `modelData` (as a `T`) into the `ModelClass` constructor.

Reply:
Gotcha.  That is a strange design decision, I think.  lodash is one of the most helpful libraries out there, but they committed to a lot of things and stuck by them.  I do like that they reworked stuff into the FP library, but they probably should have overhauled the entire thing.

Reply:
Yeah, https://ramdajs.com/ looks like an interesting alternative to lodash fp, but it's even larger and a bit intimidating haha